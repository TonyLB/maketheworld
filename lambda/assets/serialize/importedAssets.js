import { assetDB } from "/opt/utilities/dynamoDB/index.js"

import { splitType } from '/opt/utilities/types.js'

export const importedAssetIds = async (importMap) => {
    const getScopeMap = async () => {
        const fetchScopeMap = async ({ key, asset, scopedId }) => {
            const Items = await assetDB.query({
                IndexName: 'ScopedIdIndex',
                scopedId,
                KeyConditionExpression: 'DataCategory = :dc',
                ExpressionAttributeValues: {
                    ':dc': `ASSET#${asset}`
                }
            })
            if (Items.length === 0) {
                return {}
            }
            const { AssetId } = Items[0]
            const translatedAssetId = splitType(AssetId || '')[1]
            if (translatedAssetId) {
                return {
                    [key]: translatedAssetId
                }
            }
            return {}
        }
        const fetchPromises = Object.entries(importMap)
            .map(([key, { asset, scopedId }]) => (fetchScopeMap({ key, asset, scopedId })))
        const promiseReturns = await Promise.all(fetchPromises)
        return Object.assign({}, ...promiseReturns)
    }

    const getImportTree = async () => {
        const assetsToFetch = [...(new Set(Object.values(importMap).map(({ asset }) => (asset))))]
        const fetchAssetImportTree = async (asset) => {
            const { importTree = {} } = await assetDB.getItem({
                AssetId: `ASSET#${asset}`,
                DataCategory: 'Meta::Asset',
                ProjectionFields: ['importTree']
            })
            return { [asset]: importTree }
        }
        const fetchPromises = assetsToFetch.
            map((asset) => (fetchAssetImportTree(asset)))
        
        const promiseReturns = await Promise.all(fetchPromises)
        return Object.assign({}, ...promiseReturns)
    }

    const [ scopeMap, importTree ] = await Promise.all([ getScopeMap(), getImportTree() ])
    return {
        scopeMap,
        importTree
    }
}

const assetIdsFromTreeRecurse = (importTree) => {
    const returnValue = Object.entries(importTree)
        .map(([key, nested]) => ([key, ...assetIdsFromTreeRecurse(nested)]))
        .reduce((previous, list) => ([...previous, ...list]), [])
    return returnValue
}

export const assetIdsFromTree = (importTree) => {
    return [...(new Set(assetIdsFromTreeRecurse(importTree)))]
}
