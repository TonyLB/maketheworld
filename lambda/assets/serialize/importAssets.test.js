import { jest, describe, it, expect } from '@jest/globals'

jest.mock('@tonylb/mtw-utilities/dynamoDB/index.js')
import { assetDB } from '@tonylb/mtw-utilities/dynamoDB/index.js'

import { importedAssetIds, assetIdsFromTree } from './importedAssets'

describe('importedAssetIds', () => {
    it('should return scopedIds and aggregate of import trees for dependencies', async () => {
        assetDB.query.mockImplementation(async ({ scopedId }) => {
            switch(scopedId) {
                case 'VORTEX': return [{ AssetId: 'ROOM#VORTEX' }]
                case 'Welcome': return [{ AssetId: 'ROOM#12345' }]
            }
        })
        assetDB.getItem.mockResolvedValue({
            importTree: {}
        })
        const output = await importedAssetIds({
            VORTEX: { asset: 'BASE', scopedId: { key: 'VORTEX' } },
            Welcome: { asset: 'BASE', scopedId: { key: 'Welcome' } }
        })
        expect(output).toMatchSnapshot()
    })

    it('should nest import trees of dependencies', async () => {
        assetDB.query.mockImplementation(async ({ scopedId }) => {
            switch(scopedId) {
                case 'VORTEX': return [{ AssetId: 'ROOM#VORTEX' }]
                case 'Welcome': return [{ AssetId: 'ROOM#12345' }]
            }
        })
        assetDB.getItem.mockImplementation(async ({ AssetId }) => {
            switch(AssetId) {
                case 'ASSET#LayerA': 
                    return {
                        importTree: {
                            BASE: {}
                        }
                    }
                case 'ASSET#BASE': return {
                        importTree: {}
                    }
            }
        })
        const output = await importedAssetIds({
            VORTEX: { asset: 'BASE', scopedId: { key: 'VORTEX' } },
            Welcome: { asset: 'LayerA', scopedId: { key: 'Welcome' } }
        })
        expect(output).toMatchSnapshot()
    })

})

describe('assetIdsFromTree', () => {
    it('should return empty list from empty tree', () => {
        expect(assetIdsFromTree({})).toEqual([])
    })

    it('should parse a nested tree', () => {
        expect(assetIdsFromTree({
            MixedLayerA: {
                LayerA: {
                    BASE: {}
                },
                LayerB: {
                    BASE: {}
                }
            }
        })).toEqual([
            'MixedLayerA',
            'LayerA',
            'BASE',
            'LayerB'
        ])
    })
})