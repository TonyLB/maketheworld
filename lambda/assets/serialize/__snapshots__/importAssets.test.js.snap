// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importedAssetIds should nest import trees of dependencies 1`] = `
Object {
  "importTree": Object {
    "BASE": Object {},
    "LayerA": Object {
      "BASE": Object {},
    },
  },
  "scopeMap": Object {
    "VORTEX": "ROOM#VORTEX",
    "Welcome": "ROOM#12345",
  },
}
`;

exports[`importedAssetIds should return scopedIds and aggregate of import trees for dependencies 1`] = `
Object {
  "importTree": Object {
    "BASE": Object {},
  },
  "scopeMap": Object {
    "VORTEX": "ROOM#VORTEX",
    "Welcome": "ROOM#12345",
  },
}
`;
